### PR TITLE
Release v5.4.8

### PR DESCRIPTION
**Pull request by parhumm**
## Release v5.4.8

This release fixes remaining tracking issues from the 5.4.x upgrade cycle. If you upgraded from 5.3.x through 5.4.0-5.4.6, this update restores session cookies and client-side tracking automatically.

### Migration & Tracking
- **Fix:** Session cookies now restored for all upgrade paths, not just GDPR-disabled sites
- **Fix:** Client-side (JavaScript) tracking restored unconditionally — fixes zero tracking on cached sites
- **Fix:** Migration forced-resets gated to run once, preserving admin choices on future updates
- **Fix:** FingerprintJS v4 now generates fingerprints correctly — `.get()` call was missing since v3→v4 migration
- **Fix:** JS consent check now mirrors PHP logic when SlimStat banner is off

### External Database Addon
- **Fix:** Charts and reports now query the correct database for External DB addon users
- **Fix:** Real-time analytics queries use the correct database connection
- **Fix:** Complex report queries (e.g. Recent Events) now work with External DB addon
- **Fix:** Filter dropdown autocomplete queries the correct database
- **Fix:** Visit counter seeds correctly from external database for Pro addon users
- **Fix:** Index existence check uses the correct database when Custom DB is active
- **Fix:** Table prefix set correctly on custom wpdb instances

### Reports & Filters
- **Fix:** Country percentages exceeding 100% in Audience Location map — query cache now stays fresh for live date ranges (#270)
- **Fix:** Filter removal via red cross button not working
- **Fix:** Outbound Link, Notes, and Category filter dropdowns now show individual values instead of raw concatenated strings
- **Fix:** Filter 'equals' operator now works on Outbound Link, Notes, and Category columns (auto-upgraded to substring match)
- **Fix:** Chart granularity selection (Daily/Weekly/Monthly) persists across page reloads (#265)
- **Fix:** Chart granularity now syncs across all charts on the same page
- **Fix:** Chart timezone offset corrected for non-UTC servers

### Browscap Library
- **Fix:** Browscap now initializes WordPress filesystem before extraction — resolves toggle revert (#14843)
- **Fix:** Browscap errors show specific failure details instead of generic messages
- **Fix:** Downloaded Browscap files validated as ZIP before extraction
- **Fix:** Browscap download compatible with hosts that block GitHub redirects

### Improvements
- Chart granularity persisted via localStorage for cross-session consistency
- sessionStorage access wrapped in try/catch for private browsing compatibility

### CODE REVIEW

**Comment by parhumm** [comment]
## Human QA Checklist — FingerprintJS v4 Fix (34f88dc7)

### Root Cause
`FingerprintJS.load()` returns `Promise<Agent>`, not the fingerprint result. The required `agent.get()` call was missing since the v3→v4 migration — `fingerprintHash` was always empty, `&fh=` sent blank in every tracking request.

**Fix:** 3-line change in `wp-slimstat.js:1618-1620` — added `.then(fp => fp.get())` before `.then(result => ...)`.

### Automated Test Results
- **4/4 new E2E tests pass** (`fingerprint-generation.spec.ts`)
- **194/194 full E2E suite pass** — zero regressions

---

### Required Settings to Reproduce User's Configuration

Before testing, verify these settings in **WP Admin → SlimStat → Settings**:

| Setting | Location in Admin | Required Value |
|---------|-------------------|----------------|
| **Tracking Mode** | Settings tab → **General** → Tracker section | Toggle set to **Client** (not Server) |
| **Tracking Request Method** | Settings tab → **General** → Tracker section | **Admin-AJAX** (default) or REST API |
| **GDPR Compliance Mode** | Settings tab → **Tracker** → Consent Management section | **Off** |
| **Anonymize IP Addresses** | Settings tab → **Tracker** → Data Protection section | **Off** |
| **Set Tracking Cookie** | Settings tab → **Tracker** → Data Protection section | **On** |

> **Note:** `Anonymous Tracking` is a hidden setting (not visible in admin UI, default: Off). No action needed — it's off by default after 5.4.7 migration.

These are the defaults after a clean 5.4.7 migration. If any differ, the test results may vary.

---

### Step 1: Reproduce Bug (before deploying fix)

1. Open site in **incognito/private window** (not logged in)
2. Open DevTools → **Network** tab
3. **Filter by:** `admin-ajax.php` (default tracking method is Admin-AJAX)
   - If Tracking Request Method is set to REST API, filter by `slimstat/v1/hit` instead
   - If set to Ad-Blocker Bypass, filter by `.js` (the request URL looks like `/{32-char-hash}.js/`)
4. Visit any page — wait 2-3 seconds for the tracking request to appear
5. Find the **`POST admin-ajax.php`** request where the **Payload** contains `action=slimtrack`
   - ⚠️ Ignore other `admin-ajax.php` requests (WordPress makes many AJAX calls) — look specifically for the one with `action=slimtrack` in the payload
6. Click the request → **Payload** tab (Chrome) or **Request** tab (Firefox) → look for the `fh=` parameter in the form data

**Bug confirmed (before fix) — `fh=` is empty:**
```
action=slimtrack&ref=...&res=...&sw=1920&sh=1080&bw=1440&bh=900&sl=0&pp=0&fh=&tz=-210
                                                                              ^^^
                                                                         fh is EMPTY
```

**Fixed (after fix) — `fh=` has a 32-char hex value:**
```
action=slimtrack&ref=...&res=...&sw=1920&sh=1080&bw=1440&bh=900&sl=0&pp=0&fh=afdd9e9bed40bafc252f3bc7be5f15d4&tz=-210
                                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                                         32-char hex fingerprint
```

---

### Step 2: Deploy & Clear Caches

1. Update plugin files
2. Clear: browser cache, caching plugin (WP Rocket/W3TC/LiteSpeed), CDN
3. If using ad-blocker bypass: Settings → General → Advanced Options → Flush Rewrite Rules

### Step 3: Verify Fix — Browser Network Tab

- [ ] Visit any page → `POST admin-ajax.php` with `action=slimtrack` → `fh=` has **32-char hex value**
- [ ] Visit a **second page** → `fh=` value is **identical** (stable fingerprint per browser)
- [ ] Open a **different browser** → `fh=` value is **different** (unique per browser)

### Step 4: Verify Fix — Admin Dashboard

#### What the Access Log looks like for pageviews tracked DURING the broken period (5.4.0–5.4.6):

> Since this fix is already on the `development` branch, you can only see the broken state by looking at **old records** tracked before the fix was deployed. Any new pageview on this branch will already have fingerprints.

- Old anonymous visitor rows show: IP address, page URL, browser/OS info
- **No fingerprint code** next to those old pageview records
- Old pageviews from the same anonymous browser appear as **separate, unconnected visits**
- The `fingerprint` database column is `NULL` for records tracked during the broken period

#### What the Access Log looks like for NEW pageviews (after deploying the fix):
- [ ] Go to **SlimStat → Access Log**
- [ ] Each anonymous visitor row now shows a clickable **8-character fingerprint code** (e.g., `a1b2c3d4`) in a monospace `<code>` tag, displayed next to the screen resolution info
- [ ] **Click the fingerprint code** → the Access Log filters to show **all pageviews from that same browser**
- [ ] **Multiple pageviews** from the same anonymous browser are **grouped under one visit session** with a shared visit header, instead of appearing as separate disconnected visitors
- [ ] Go to **SlimStat → Overview** → the "Unique Visitors" count is now **more accurate** — only truly first-time browsers are tagged as new visitors

#### Database verification (optional):
```sql
-- Before fix: fingerprint is NULL for anonymous visitors
SELECT id, fingerprint FROM wp_slim_stats ORDER BY id DESC LIMIT 10;
-- Result: all fingerprint values are NULL

-- After fix: fingerprint has 32-char hex values
SELECT id, fingerprint FROM wp_slim_stats ORDER BY id DESC LIMIT 10;
-- Result: e.g., afdd9e9bed40bafc252f3bc7be5f15d4
```

---

### Step 5: Verify Privacy Gates Still Work

| Setting to Change | Where in Admin | How to Test | Expected in Network Tab (`fh=`) | Expected in Access Log |
|-------------------|----------------|-------------|--------------------------------|----------------------|
| **GDPR Compliance Mode → On** | Tracker → Consent Management | Enable GDPR, visit as new anonymous user (don't click accept on the consent banner) | `fh=` is **empty** | No fingerprint code in Access Log rows |
| **Anonymize IP → On** | Tracker → Data Protection | Enable Anonymize IP Addresses, visit a page | `fh=` still has a 32-char value (browser sends it) | But fingerprint column is **NULL** in DB — server-side gate discards it silently |
| **Anonymous Tracking → On** (GDPR Off) | Hidden setting — requires DB or WP-CLI access (not in admin UI). Already covered by automated E2E test. | `fh=` is **empty** — JS intentionally skips fingerprinting | No fingerprint code in Access Log rows |

### Step 6: No Breakage

- [ ] Logged-in admin visits still tracked normally in Access Log
- [ ] No JavaScript errors in browser console (DevTools → Console tab)
- [ ] Charts and reports load without errors on all SlimStat pages
- [ ] Page load speed not noticeably affected (FingerprintJS runs via `requestIdleCallback`, non-blocking)

---

### Summary: What Changes for End Users

> **Important:** This fix is already merged into the `development` branch. If you're testing on this branch, all **new** pageviews will already have fingerprints. The "Production (broken)" column describes what users on published versions 5.4.0–5.4.6 currently experience. **Old pageview records** tracked during the broken period will still show NULL fingerprints — only new visits after deploying get fingerprints.

| Area | Production 5.4.0–5.4.6 (broken) | After Deploying This PR |
|------|----------------------------------|------------------------|
| **Access Log rows** | No fingerprint code for anonymous visitors | Clickable `a1b2c3d4` fingerprint code per visitor |
| **Visitor grouping** | Each pageview appears as a separate unconnected visit | Same-browser pageviews grouped under one visit session |
| **"New Visitors" metric** | Inflated — every anonymous hit looks "new" | Accurate — only first-time browsers tagged new |
| **Session detection** | Falls back to IP + User-Agent hashing (unreliable behind NAT/VPN) | Fingerprint-based matching (unique per browser) |
| **Fingerprint DB column** | `NULL` for anonymous visitors tracked during broken period | 32-char hex value for new visits (e.g., `afdd9e9bed40bafc252f3bc7be5f15d4`) |
| **Network tab `fh=` param** | Empty (`&fh=&`) | 32-char hex value |

**Comment by parhumm** [comment]
## Updated QA Checklist — Browscap WP_Filesystem Fix (#14843)

### How to Reproduce Bob's Bug

**Required settings state (before testing):**
1. Go to **SlimStat → Settings → Tracker** tab (tab 2)
2. Ensure **Browscap Library** toggle is **OFF**
3. No special `wp-config.php` changes needed — default WordPress install
4. No `FS_METHOD` constant defined (this is the default on most hosts)
5. Server must be able to reach `github.com` (for Browscap database download)

**Reproduce the bug (on v5.4.6 and earlier v5.4.7 builds before the WP_Filesystem fix):**
1. Toggle Browscap Library **ON**
2. Click **Save**
3. **Result with bug:**
   - Toggle silently reverts to **OFF**
   - v5.4.6 error message: *"There was an error uncompressing the Browscap data file on your server. Please check your folder permissions and PHP configuration."* (generic, unhelpful)
   - v5.4.7 (before this fix) error message: *"There was an error uncompressing the Browscap data file: Could not access filesystem."* (more specific but still fails)
4. Refreshing the page confirms toggle is OFF — setting was never saved

**Root cause:** WordPress's `unzip_file()` requires `WP_Filesystem()` to be called first. SlimStat never called it. On servers where no other plugin initializes the filesystem API during that request, `unzip_file()` immediately fails with "Could not access filesystem."

---

### What the Admin Sees — Before Fix vs After Fix

#### Before fix (toggle reverts — what Bob experienced):

| Admin Panel Area | What you see |
|-----------------|-------------|
| **Settings → Tracker** | Browscap toggle always reverts to OFF after save |
| **Error message (v5.4.6)** | *"error uncompressing... check folder permissions and PHP configuration"* |
| **Error message (v5.4.7 pre-fix)** | *"error uncompressing... Could not access filesystem."* |
| **Reports → Browsers** | Browser names still appear (e.g., "Chrome", "Firefox") — **UADetector fallback works** |
| **Reports → Browser Type** | Basic detection — shows Human/Bot/Mobile via UADetector |
| **Reports → Platforms** | OS detection still works (e.g., "Windows", "Linux") via UADetector |
| **Tracking** | **Fully functional** — visits are recorded normally. Only browser detection detail is reduced |

#### After fix (toggle stays ON):

| Admin Panel Area | What you see |
|-----------------|-------------|
| **Settings → Tracker** | Browscap toggle stays **ON** after save ✓ |
| **Success message** | *"The Browscap data file has been installed on your server."* |
| **Reports → Browsers** | More detailed browser names with precise version numbers (e.g., "Chrome 120.0" vs just "Chrome") |
| **Reports → Browser Type** | More accurate classification: Human (0), Bot/Crawler (1), Mobile Device (2), Syndication Reader (3) — Browscap better distinguishes tablets, touch devices, and crawlers |
| **Reports → Platforms** | More precise OS detection via Browscap's database |
| **Tracking** | Identical — same visits recorded, richer browser/device metadata |

#### If server genuinely can't initialize filesystem (after fix):

| Admin Panel Area | What you see |
|-----------------|-------------|
| **Settings → Tracker** | Toggle reverts to OFF (correct — can't install the library) |
| **Error message** | **Actionable:** *"Could not initialize the WordPress filesystem. Please check your server permissions or set FS_METHOD to 'direct' in wp-config.php."* |
| **Existing Browscap cache** | **Preserved** — old cache is NOT deleted if filesystem init fails (safe rollback) |
| **Reports** | UADetector fallback works normally — no data loss |

**Key point:** Browscap is an **optional enhancement** for browser detection. When the toggle is OFF (bug or by choice), SlimStat uses its built-in UADetector and everything works — you just get less precise browser/device info in reports. No tracking data is lost either way.

---

### Test Steps

#### Test 1: Browscap Toggle ON (the fix)
1. **SlimStat → Settings → Tracker** tab
2. Toggle Browscap Library **ON** → **Save**
3. Check:
   - [ ] Success message: *"The Browscap data file has been installed on your server."*
   - [ ] No error about "filesystem" or "uncompressing"
   - [ ] Toggle stays **ON** after page reload
   - [ ] Navigate away and back — still ON

#### Test 2: Browscap Toggle OFF
1. Toggle Browscap **OFF** → **Save**
2. Check:
   - [ ] Message: *"The Browscap data file has been uninstalled from your server."*
   - [ ] Toggle stays **OFF** after reload

#### Test 3: Other settings not clobbered
1. Toggle Browscap **ON** AND change **Session Duration** (e.g., to 3600)
2. Click **Save**
3. Check:
   - [ ] Browscap is ON
   - [ ] Session Duration saved the new value
   - [ ] No other settings reverted

#### Test 4: Verify browser detection in reports
1. With Browscap **ON**, visit the frontend site from a browser
2. Go to **SlimStat → Activity Log** or **Right Now** report
3. Check:
   - [ ] Browser name and version are shown (e.g., "Chrome 120.0")
   - [ ] Platform is detected (e.g., "win10")
   - [ ] Browser type icon is accurate (Human/Bot/Mobile)

**Comment by parhumm** [comment]
## New in this push: Fix for country percentage >100% (#270)

### Bug
The Audience Location map (`slim_p6_01`) showed country percentages exceeding 100% (e.g., UK at 110.94%).

### Root Cause
Cache asymmetry between `Query::getVar()` and `Query::getAll()`:
- `$pageviews` (via `count_records()` → `getVar()`) cached the full date range for up to 1 hour **without date splitting**
- Per-country counts (via `get_top()` → `getAll()`) split into cached historical + **always-fresh live** data
- Stale denominator < fresh numerator → percentages >100%

### Fix
`getVar()`, `getRow()`, `getCol()` now call `getSplitDateRanges()` and skip cache when the date range includes today — consistent with how `getAll()` already handles live data.

### Impact Analysis
| Method | Callers with caching | Risk |
|--------|---------------------|------|
| `getVar()` | 10 (admin aggregates) | LOW — fresher data, no behavior change |
| `getRow()` | 0 | NONE |
| `getCol()` | 0 (unused) | NONE |

Historical-only date ranges still cached (no performance regression).

### Tests Added
- **Unit**: `QueryGetVarCacheTest.php` — 3 tests (cache for past ranges, skip cache for live ranges, no cache when disabled)
- **E2E**: `country-percentage-cache-fix.spec.ts` — 8 scenarios (basic regression, cache freshness, historical caching preserved, map vs list consistency, browser reports, sidebar pageviews, async_load, custom DB prefix)

### Commits
- `7ee194c2` — fix: skip cache for getVar/getRow/getCol when date range includes today
- `c4dd1b90` — test(e2e): 8 regression tests for country percentage cache fix
- `8daadfe6` — fix: unit test prepare mock + changelog entry

CI running: waiting for Tier 1 (PHPUnit) + Tier 2 (E2E) results.

**Comment by parhumm** [comment]
## CI Results — #270 cache fix (all green)

| Job | PHP | Status |
|-----|-----|--------|
| Tier 1 · PHPUnit | 7.4 | :white_check_mark: Pass |
| Tier 1 · PHPUnit | 8.2 | :white_check_mark: Pass |

**66 unit tests, 105 assertions** — all passing including 3 new regression tests for `getVar()` cache behavior.

### Additional fix: auto-clear stale cache on upgrade

Added transient cleanup in `update_tables_and_options()` so existing sites don't need to manually click "Clear Cache" after updating. Only deletes `_transient_wp_slimstat_cache_*` entries (query cache), does not touch other SlimStat or WordPress transients.
